### ISSUE TITLE
--nv flag does not work in apptainer nested inside apptainer

### ISSUE THREAD

**Issue by DrDaveD** [apptainer/apptainer#464]
This is a continuation of apptainer/singularity#5759.

**Comment by luator** [comment]
I tried a few thinks now.  The [workaround](https://github.com/apptainer/singularity/issues/5759#issuecomment-919523970) does work for me, when I pull `docker://matterminers/wlcg-wn\:latest`, run it with Apptainer and then use Singularity that is installed inside, i.e.

workaround.sh:
```
TMPD=`mktemp -d`
(echo '#!/bin/bash'; echo 'exec /usr/sbin/ldconfig -C '"$TMPD"'/ld.so.cache "$@"') >$TMPD/ldconfig
chmod +x $TMPD/ldconfig
PATH=$TMPD:$PATH
ldconfig $LD_LIBRARY_PATH
```
Commands:
```
$ apptainer shell --nv wlcg-wn_latest.sif
> . workaround.sh
> singularity shell --nv wlcg-wn_latest.sif
> > nvidia-smi
```
--> good.

In the other issue, you said "I bind-mounted /cvmfs" to get Apptainer into the container.  However, I don't have `cvmfs`, so not sure how I could do this.

Instead I tried with a more minimal example, using an Ubuntu base image and only install Apptainer:
```
Bootstrap: docker
From: ubuntu:20.04

%post
    export DEBIAN_FRONTEND=noninteractive
    apt-get update

    apt-get install -y wget 
    wget https://github.com/apptainer/apptainer/releases/download/v1.0.2/apptainer_1.0.2_amd64.deb

    apt-get install -y ./apptainer_1.0.2_amd64.deb
```
(alternatively build Apptainer from source with `./mconfig --without-suid`, so no sudo is needed when running).

Unfortunately, this does not work:
```
$ sudo apptainer shell --nv --bind=$(pwd) --pwd=$(pwd) ubuntu_with_apptainer.sif
> . workaround.sh
> apptainer shell --nv ubuntu_with_apptainer.sif
> > nvidia-smi
```
results in
> WARNING: Could not find any nv libraries on this host!

and

> NVIDIA-SMI couldn't find libnvidia-ml.so library in your system. Please make sure that the NVIDIA Display Driver is properly installed and present in your system.
Please also try adding directory that contains libnvidia-ml.so to your system PATH.

Any advice would be very welcome.

**Comment by luator** [comment]
I also tried creating an image based on wlcg-wn_latest.sif and install Apptainer in it (from source).  The build succeeded but when running it, it doesn't even find the `nvidia-smi` executable in the inner container:
```
$ apptainer shell --nv wlcg_with_apptainer.sif      
Apptainer> . workaround.sh 
Apptainer> apptainer shell --nv wlcg_with_apptainer.sif
WARNING: While finding nv bind points: could not retrieve ld cache: could not execute ldconfig: exit status 1
WARNING: Could not find any nv libraries on this host!
INFO:    Converting SIF file to temporary sandbox...
Apptainer> nvidia-smi
bash: nvidia-smi: command not found
```

This is the definition file to create the image:
```
bootstrap: localimage
from: wlcg-wn_latest.sif

%post
    mkdir /tmp_build
    cd /tmp_build

    # install go
    export GOVERSION=1.18.2 OS=linux ARCH=amd64

    wget -O /tmp/go${GOVERSION}.${OS}-${ARCH}.tar.gz \
        https://dl.google.com/go/go${GOVERSION}.${OS}-${ARCH}.tar.gz
    tar -C /usr/local -xzf /tmp/go${GOVERSION}.${OS}-${ARCH}.tar.gz

    # get apptainer code
    APPTAINER_VERSION=1.0.2
    wget https://github.com/apptainer/apptainer/releases/download/v${APPTAINER_VERSION}/apptainer-${APPTAINER_VERSION}.tar.gz
    tar -xzf apptainer-${APPTAINER_VERSION}.tar.gz

    # build and install
    export PATH=$PATH:/usr/local/go/bin
    cd apptainer-${APPTAINER_VERSION}
    ./mconfig --without-suid
    cd builddir
    make
    make install

    # delete build files
    cd /
    rm -rf /tmp_build
```

**Comment by DrDaveD** [comment]
I see, you are right.  The problem was introduced in apptainer 1.0.0 via apptainer/singularity#6224 which was imported from a Sylabs PR.  There the ldconfig path was put into the configuration file at compile time rather than searching $PATH at run time.  The problem must also exist in singularity-ce >= 3.9.0 and greater.

The good news is that this is already fixed in the main branch via #367 and so will be in 1.1.0.  It goes back to searching $PATH by default and includes the user's $PATH when running unprivileged.  I verified it by installing a package built from the current main branch in my image and changing the configuration to set `allow setuid = no` with this line:
```
sed -i 's/^allow setuid = yes/allow setuid = no/' /etc/apptainer/apptainer.conf
```
Still, it would be nicer of course if the workaround was not needed.

Another way to make it work even with apptainer 1.0.x should be the following:
```
export APPTAINER_BINDPATH=/.singularity.d/libs
```
That doesn't get rid of the `Could not find any nv libraries on this host!` warning but it does put the libraries into $LD_LIBRARY_PATH where they should be visible to nvidia-smi.  It shouldn't be hard to make a PR that also looks for the libraries in /.singularity.d/libs (or $LD_LIBRARY_PATH to be more general) and uses them instead of the original paths.

> I don't have cvmfs, so not sure how I could do this.

cvmfs can be installed by following instructions on its [home page](https://cernvm.cern.ch/fs/).

**Comment by luator** [comment]
Thanks a lot for looking into this!  I can confirm that setting `export APPTAINER_BINDPATH=/.singularity.d/libs` works.
